flink-ml-core/src/main/java/org/apache/flink/ml/common/iteration/TerminateOnMaxIterOrTol.java

```diff
@@ -16,29 +16,53 @@
  * limitations under the License.
  */
 
-package org.apache.flink.test.iteration.operators;
+package org.apache.flink.ml.common.iteration;
 
 import org.apache.flink.api.common.functions.FlatMapFunction;
 import org.apache.flink.iteration.IterationListener;
 import org.apache.flink.util.Collector;
 
-/** An termination criteria function that asks to stop after the specialized round. */
-public class RoundBasedTerminationCriteria
-        implements FlatMapFunction<EpochRecord, Integer>, IterationListener<Integer> {
+/**
+ * A FlatMapFunction that emits values iff the iteration's epochWatermark does not exceed a certain
+ * threshold and the loss exceeds a certain tolerance.
+ *
+ * <p>When the output of this FlatMapFunction is used as the termination criteria of an iteration
+ * body, the iteration will be executed for at most the given `maxIter` iterations. And the
+ * iteration will terminate once any input value is smaller than or equal to the given `tol`.
+ */
+public class TerminateOnMaxIterOrTol
+        implements IterationListener<Integer>, FlatMapFunction<Double, Integer> {
+
+    private final int maxIter;
 
-    private final int maxRound;
+    private final double tol;
 
-    public RoundBasedTerminationCriteria(int maxRound) {
-        this.maxRound = maxRound;
+    private double loss = Double.NEGATIVE_INFINITY;
+
+    public TerminateOnMaxIterOrTol(int maxIter, double tol) {
+        this.maxIter = maxIter;
+        this.tol = tol;
+    }
+
+    public TerminateOnMaxIterOrTol(int maxIter) {
+        this.maxIter = maxIter;
+        this.tol = Double.NEGATIVE_INFINITY;
+    }
+
+    public TerminateOnMaxIterOrTol(double tol) {
+        this.maxIter = Integer.MAX_VALUE;
+        this.tol = tol;
     }
 
     @Override
-    public void flatMap(EpochRecord integer, Collector<Integer> collector) throws Exception {}
+    public void flatMap(Double value, Collector<Integer> out) {
+        this.loss = value;
```

Comment:
What if the input value sequence may be increasing? Should we use the minimum input value observed in this epoch?

Reply:
Hmm, there should be only one loss value in each epoch.

Reply:
The issue here is that there is no rule forcing `only one loss value in each epoch`. If there should indeed be `only one loss value in each epoch`, could we throw exception in `flatMap()` to enforce it?

I think it is better not to limit the number of input values in each epoch so that this operator can also be used in asynchronous iteration. And if we agree to do so, we probably need to use the minimum value here.

Reply:
I think throwing exception to enforce only one input in each epoch is a viable solution.

Users are not supposed to use `TerminateOnMaxIterOrTol` or `TerminateOnMaxIter` for asynchronous iterations --- Users are expected to decide termination relying on their own loss..

Reply:
Sounds good. Let's throw exception in this case.